torch/_export/__init__.py

```diff
@@ -785,48 +911,48 @@ def _export(
     if isinstance(f, torch.nn.Module):
         _normalize_nn_module_stack(gm_torch_level, type(f))
 
-    aot_export_args = (*fake_args, *_reorder_kwargs_by_names(orig_args, fake_args, fake_kwargs).values())
-
-    user_input_names = _prepare_module(gm_torch_level, aot_export_args)
-
-    # Note: aot_export_module doesn't accept kwargs, we'd like to reorder the kwargs as an OrderedDict
-    # to follow the order in orig_args and correctly call gm_torch_level
-
-    # This _reparametrize_module makes sure inputs and gm_torch_level.params/buffers have the same fake_mode,
-    # otherwise aot_export_module will error out because it sees a mix of fake_modes.
-    # And we want aot_export_module to use the fake_tensor mode in dynamo to keep the pipeline easy to reason about.
-    with torch.nn.utils.stateless._reparametrize_module(gm_torch_level, fake_params_buffers):
-        gm, graph_signature = aot_export_module(
-            gm_torch_level,
-            (),
-            trace_joint=False
-        )
-    user_inputs_to_mutate = _unwrap_user_inputs(gm, graph_signature, user_input_names)
-
-    def to_str_list(sig_component: List[Any]):
-        return [str(v) for v in sig_component]
+    def _process_user_inputs(aot_export):
+        def _aot_export_strict(gm_torch_level: torch.fx.GraphModule, args, **kwargs):
+            user_input_names = _unlift_user_inputs_to_buffers(gm_torch_level, args)
+            gm, graph_signature = aot_export(gm_torch_level, (), **kwargs)
+            user_inputs_to_mutate = _lift_buffers_to_user_inputs(gm, graph_signature, user_input_names)
+            # TODO unfortunately preserving graph-level metadata is not
+            # working well with aot_export. So we manually copy it.
+            # (The node-level meta is addressed above.)
+            gm.meta.update(gm_torch_level.meta)
+            assert "user_inputs_to_mutate" not in gm.meta
+            gm.meta["user_inputs_to_mutate"] = user_inputs_to_mutate
+            return gm, graph_signature
+
+        return _aot_export_strict
+
+    ep_non_strict = _export_non_strict(
+        gm_torch_level,
+        fake_args,
+        _reorder_kwargs_by_names(orig_args, fake_args, fake_kwargs),
+        fake_params_buffers,
+        transform=_process_user_inputs
```

Comment:
Non-blocking: do we really need the transform kwarg? It looks like wrapper. When seeing the word transform, I though it's some passes over the graph module. Can we do the wrapping eagerly then pass the wrapped into _export_non_strict? It's a bit more readable I guess.

Reply:
yeah good point. I might remove this at some point but right now I need this to support user input mutations which requires some code to run before and after aot_export_module, so it has a sandwich structure. I can add a TODO here.